lib/pundit/context.rb

```diff
@@ -108,7 +108,7 @@ def policy(record)
     # @raise [InvalidConstructorError] if the policy constructor called incorrectly
     # @return [Object] instance of policy class with query methods
     def policy!(record)
-      policy = policy_finder(record).policy!
+      policy = policy_cache[record] || policy_finder(record).policy!
```

Comment:
If you agree with this mentality, I wonder if this could be taken a step further by moving this or similar logic into the `def policy_finder` private method.